server/src/main/java/org/apache/druid/server/http/RulesResource.java

```diff
@@ -181,4 +186,37 @@ private List<AuditEntry> getRuleHistory(
     return auditManager.fetchAuditHistory(AUDIT_HISTORY_TYPE, theInterval);
   }
 
+  /**
+   * Validate rules. Throws an exception if a rule contain an interval that will overshadow another rules' interval.
+   * Rules that will be evaluated at some point are considered to be non-overshadowing.
+   * @param rules Datasource rules.
+   */
+  private void validateRules(final List<Rule> rules)
```

Comment:
The logic here can be error prone and would require thorough testing. Best move it out into a separate class called `RulesValidator` or as a public static method in `Rules` utility class itself.

Reply:
Good idea. Moved it to `Rules` class.